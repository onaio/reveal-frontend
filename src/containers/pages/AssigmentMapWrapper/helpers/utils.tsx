import { Dictionary } from '@onaio/utils';
import { Feature } from '@turf/turf';
import { History } from 'history';
import { EventData, Map } from 'mapbox-gl';
import { AssignmentMapWrapperProps } from '..';
import { PLAN_TYPES_WITH_MULTI_JURISDICTIONS } from '../../../../configs/env';
import { getPlanType } from '../../../../helpers/utils';
import { SELECTION_REASON } from '../../../../store/ducks/opensrp/hierarchies/constants';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/**
 * Handler to get current parent id from clicked jurisdiction on map
 * @param props - AssignmentMapWrapper component props
 * @param setMapParent - function to set map parent id to state
 * @param history - history session object
 */

export const onJurisdictionClick = (
  props: AssignmentMapWrapperProps,
  setMapParent: (currentId: string) => void,
  history: History
) => {
  function handleJurisdictionClick(_: Map, e: EventData) {
    // Destructure event data
    const { point, target, originalEvent } = e;
    const {
      currentParentNode,
      rootJurisdictionId,
      selectNodeCreator,
      deselectNodeCreator,
      currentChildren,
      plan,
      baseAssignmentURL,
    } = props;
    // grab underlying features from map
    const features: Feature[] = target.queryRenderedFeatures(point);
    if (!features.length) {
      return false;
    }

    for (const feature of features) {
      const activeJurisdictionId =
        (feature && feature.id) ||
        (feature && feature.properties && feature.properties.jurisdiction_id);
      if (activeJurisdictionId) {
        const currentId = activeJurisdictionId;
        let activeCurrentNode: TreeNode | undefined;

        if (currentParentNode) {
          // ensure currentparentnode is not undefined first
          // Handle selection for admin level 0
          if (
            currentParentNode.model.id === rootJurisdictionId &&
            currentId === currentParentNode.model.id
          ) {
            activeCurrentNode = currentParentNode;
          } else {
            // Handle selection for admin level > 0
            activeCurrentNode = currentChildren.find(
              (node: TreeNode) => node.model.id === currentId
            );
          }
        }
        const thePlanType = getPlanType(plan);
        if (
          activeCurrentNode &&
          originalEvent.altKey &&
          PLAN_TYPES_WITH_MULTI_JURISDICTIONS.includes(thePlanType)
        ) {
          if (!nodeIsSelected(activeCurrentNode)) {
            selectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          } else {
            deselectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          }
          setMapParent(currentId);
        } else {
          if (activeCurrentNode && activeCurrentNode.hasChildren()) {
            history.push(`${baseAssignmentURL}/${currentId}`);
          }
        }
        break;
      }
    }
  }
  return handleJurisdictionClick;
};

/**
 * Map component popup handler
 * @param map - mapbox map instance
 * @param event - map event object
 */

export const buildMouseMoveHandler = (map: Map, event: EventData) => {
  const popup = new mapboxgl.Popup({
    closeButton: false,
  });
  // Added this part to remove duplicate popup containers showing on map when hovering
  const nodes: any = document.getElementsByClassName('mapboxgl-popup');
  let node: HTMLDivElement;
  if (nodes && nodes.length) {
    for (node of nodes) {
      if (node.style) {
        node.style.display = 'none';
      }
    }
  }
  let content: string = '';
  let feature: Feature | any = {};
  popup.remove();
  // grab underlying features from map
  const features: any = event.target.queryRenderedFeatures(event.point) as Dictionary[];
  //
  if (!features.length) {
    return;
  }

  // Change the cursor style to pointer if features exist.
  map.getCanvas().style.cursor = features.length ? 'pointer' : 'grab';

  // loop through features, setting the name property to content var
  for (feature of features) {
    if (feature && feature.properties && feature.properties.name) {
      content += `<div class="jurisdiction-name"><center>${feature.properties.name}</center></div>`;
    }
    /**
     * Break out of loop once content is set
     * This helps in a case where we could have 2 or more identical features
     */
    if (content) {
      break;
    }
  }
  /**
   * Add popup to map after content is set
   */
  if (content.length) {
    popup
      .setLngLat(map.unproject(event.point))
      .setHTML(content)
      .addTo(map);
  } else {
    // clear existing popups if content is empty
    popup.remove();
    return;
  }
  return true;
};
