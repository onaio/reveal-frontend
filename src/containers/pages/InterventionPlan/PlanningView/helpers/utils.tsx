import { DrillDownColumn } from '@onaio/drill-down-table/dist/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell } from 'react-table';
import { DATE_CREATED, NAME, STATUS_HEADER } from '../../../../../configs/lang';
import { planStatusDisplay } from '../../../../../configs/settings';
import { ASSIGN_JURISDICTIONS_URL, DRAFT_IRS_PLAN_URL } from '../../../../../constants';
import { PlanRecord } from '../../../../../store/ducks/plans';

/** Columns definition for IRS drafts page table */
export const commonColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Header: DATE_CREATED,
    accessor: 'plan_date',
    maxWidth: 50,
    minWidth: 20,
  },
  {
    Header: STATUS_HEADER,
    accessor: (d: PlanRecord) => planStatusDisplay[d.plan_status] || d.plan_status,
    id: 'plan_status',
    maxWidth: 50,
    minWidth: 20,
  },
];

export const irsDraftPageColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <div>
          <Link to={`${DRAFT_IRS_PLAN_URL}/${original.id || original.plan_id}`}>{cell.value}</Link>
        </div>
      );
    },
    Header: NAME,
    accessor: 'plan_title',
    minWidth: 200,
  },
  ...commonColumns,
];

export const draftPageColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <div>
          <Link to={`${ASSIGN_JURISDICTIONS_URL}/${original.id || original.plan_id}`}>
            {cell.value}
          </Link>
        </div>
      );
    },
    Header: NAME,
    accessor: 'plan_title',
    minWidth: 200,
  },
  ...commonColumns,
];
