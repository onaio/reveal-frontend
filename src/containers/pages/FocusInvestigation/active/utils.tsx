import React from 'react';
import { RouteComponentProps } from 'react-router';
import { Column } from 'react-table';
import DrillDownTableLinkedCell from '../../../../components/DrillDownTableLinkedCell';
import {
  defaultOptions,
  renderInFilterFactory,
} from '../../../../components/Table/DrillDownFilters/utils';
import { NoDataComponent } from '../../../../components/Table/NoDataComponent';
import { NO_INVESTIGATIONS_FOUND } from '../../../../configs/lang';
import { extractPlan, removeNullJurisdictionPlans } from '../../../../helpers/utils';
import { OpenSRPService } from '../../../../services/opensrp';
import { Plan } from '../../../../store/ducks/plans';

/** construct props for DrillDown for activeFI page
 * Generic type D is the type of data that will be passed to the Table
 * its not `Plan` because the plan goes through a few transformations before being rendered,
 * at render its structure will have mutated to a Dictionary courtesy of 'extractPlan'
 * @param columns - react-table compatible columns
 * @param data - array of data to be rendered in table
 * @param queryParam - the key to be used by searchForm to use as key in searchParam
 */
export const createTableProps = <D extends object, TProps extends RouteComponentProps>(
  columns: Array<Column<D>>,
  data: Plan[] | null,
  props: TProps,
  paramString: string,
  service: typeof OpenSRPService,
  noDataMessage: string = NO_INVESTIGATIONS_FOUND
) => {
  const cleanedData = data !== null ? data : [];
  const jurisdictionValidPlans = removeNullJurisdictionPlans(cleanedData);
  const thePlans = jurisdictionValidPlans.map(extractPlan);

  return {
    CellComponent: DrillDownTableLinkedCell,
    columns,
    data: thePlans,
    renderInBottomFilterBar: renderInFilterFactory({
      showColumnHider: false,
      showFilters: false,
      showPagination: true,
      showRowHeightPicker: false,
      showSearch: false,
    }),
    renderInTopFilterBar: renderInFilterFactory({
      ...defaultOptions,
      componentProps: props,
      queryParam: paramString,
      serviceClass: service,
    }),
    renderNullDataComponent: () => <NoDataComponent message={noDataMessage} />,
    useDrillDown: false,
  };
};
