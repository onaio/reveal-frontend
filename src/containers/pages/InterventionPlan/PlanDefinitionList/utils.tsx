import { DrillDownColumn } from '@onaio/drill-down-table';
import { Dictionary } from '@onaio/utils/dist/types/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell, Row } from 'react-table';
import { isFIOrDynamicFI } from '../../../../components/forms/PlanForm/helpers';
import { DISPLAYED_PLAN_TYPES } from '../../../../configs/env';
import {
  CASE_TRIGGERED_TITLE,
  FOCUS_INVESTIGATION_STATUS_REASON,
  INTERVENTION_TYPE_LABEL,
  LAST_MODIFIED,
  ROUTINE_TITLE,
  STATUS_HEADER,
  TITLE,
} from '../../../../configs/lang';
import { planStatusDisplay, UseContext } from '../../../../configs/settings';
import {
  CASE_TRIGGERED,
  FI_REASON_CODE,
  INTERVENTION_TYPE_CODE,
  PLAN_UPDATE_URL,
  ROUTINE,
} from '../../../../constants';
import { InterventionType, PlanRecord } from '../../../../store/ducks/plans';

/**
 * sort plan UseContext values
 * @param {Row<PlanRecord>} rowA - table props
 * @param {Row<PlanRecord>} rowB - table props
 * @param {string} code - Plan use context code
 */
const sortUseContext = (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>, code: string) => {
  /** custom sort function based useContext */
  const getIntervention = (row: Row<PlanRecord>) => {
    const interventionPlanContext = (row.original as Dictionary).plan_useContext.filter(
      (context: UseContext) => context.code === code
    );
    return interventionPlanContext.length > 0 ? interventionPlanContext[0].valueCodableConcept : '';
  };
  return getIntervention(rowA) > getIntervention(rowB) ? 1 : -1;
};

/**
 *
 * @param {UseContext[]} value - List of plan use contexts
 * @param {string} code - Plan use context code
 */
const getUseContext = (value: UseContext[], code: string) => {
  const typeUseContext = value.filter((e: any) => e.code === code);
  if (typeUseContext[0]?.valueCodableConcept === CASE_TRIGGERED) {
    return CASE_TRIGGERED_TITLE;
  }
  if (typeUseContext[0]?.valueCodableConcept === ROUTINE) {
    return ROUTINE_TITLE;
  }
  return typeUseContext.length > 0 ? typeUseContext[0].valueCodableConcept : null;
};

// plan intervention type column
const interventionTypeColumn = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, INTERVENTION_TYPE_CODE),
  Header: INTERVENTION_TYPE_LABEL,
  accessor: 'plan_useContext',
  id: 'plan_use_context_intervention_type',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, INTERVENTION_TYPE_CODE),
} as DrillDownColumn<PlanRecord>;

// plan FI reason column
const fiReasonColumns = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, FI_REASON_CODE),
  Header: FOCUS_INVESTIGATION_STATUS_REASON,
  accessor: 'plan_useContext',
  id: 'plan_use_context_fi_reason',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, FI_REASON_CODE),
} as DrillDownColumn<PlanRecord>;

const displayedPlans = [...DISPLAYED_PLAN_TYPES] as InterventionType[];
const showFiReasonColumns =
  displayedPlans.length === 2 &&
  isFIOrDynamicFI(displayedPlans[0]) &&
  isFIOrDynamicFI(displayedPlans[1]);

const interventionOrFiReasonColumn = showFiReasonColumns ? fiReasonColumns : interventionTypeColumn;

export const TableColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <Link to={`${PLAN_UPDATE_URL}/${original.id}`} key={original.id}>
          {cell.value}
        </Link>
      );
    },
    Header: TITLE,
    accessor: 'plan_title',
    minWidth: 200,
  },
  interventionOrFiReasonColumn,
  {
    Cell: (cell: Cell<PlanRecord>) => {
      return planStatusDisplay[cell.value] || null;
    },
    Header: STATUS_HEADER,
    accessor: 'plan_status',
    maxWidth: 30,
    minWidth: 10,
  },
  {
    Header: LAST_MODIFIED,
    accessor: 'plan_date',
    maxWidth: 30,
    minWidth: 20,
  },
];
