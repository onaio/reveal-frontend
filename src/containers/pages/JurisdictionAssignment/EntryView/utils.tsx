/** collection of hooks that are used in the jurisdiction Assignment pages
 * these hooks majorly focus on abstracting the opensrp service calls.
 */
import { Result } from '@onaio/utils/dist/types/types';
import React from 'react';
import { ActionCreator } from 'redux';
import { getAncestors } from '../../../../components/TreeWalker/helpers';
import {
  COULD_NOT_LOAD_JURISDICTION,
  COULD_NOT_LOAD_JURISDICTION_HIERARCHY,
  COULD_NOT_LOAD_PLAN,
} from '../../../../configs/lang';
import { PlanDefinition } from '../../../../configs/settings';
import {
  loadJurisdiction,
  LoadOpenSRPHierarchy,
} from '../../../../helpers/dataLoading/jurisdictions';
import { loadOpenSRPPlan } from '../../../../helpers/dataLoading/plans';
import { StartLoading, StopLoading } from '../../../../helpers/useLoadingReducer';
import { OpenSRPService } from '../../../../services/opensrp';
import { FetchedTreeAction } from '../../../../store/ducks/opensrp/hierarchies';
import { RawOpenSRPHierarchy, TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { AddPlanDefinitionAction } from '../../../../store/ducks/opensrp/PlanDefinition';
import { HandleBrokenPage } from '../helpers/utils';

/** hook to load a plan and dispatch to store
 * @param routePlanId - id of the plan, usually got from the route
 * @param plan - if the plan exists, otherwise null
 * @param serviceClass -  opensrp service
 * @param fetchPlanCreator - action creator to be called with plan payload once fetched
 * @param handleBrokenPage - helper to handle errors
 * @param stopLoading- helper to denote when loading should stop
 * @param startLoading - helper to denote when loading should start
 */
export function usePlanEffect(
  routePlanId: string,
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  fetchPlanCreator: ActionCreator<AddPlanDefinitionAction>,
  handleBrokenPage: HandleBrokenPage,
  stopLoading: StopLoading,
  startLoading: StartLoading
) {
  React.useEffect(() => {
    // create promise to fetch the plan
    const planLoadingKey = 'planMetadata';
    startLoading(planLoadingKey, !plan);
    const planDataPromise = loadOpenSRPPlan(routePlanId, serviceClass, fetchPlanCreator).catch(
      _ => {
        handleBrokenPage(COULD_NOT_LOAD_PLAN);
      }
    );

    Promise.all([planDataPromise])
      .finally(() => {
        stopLoading(planLoadingKey);
      })
      .catch(e => handleBrokenPage(e.message));
  }, []);
}

/** gets the rootJurisdictionId - uses any of the assigned/existing jurisdictions
 * @param plan - the plan if it exists, otherwise null
 * @param serviceClass - the opensrpService
 * @param handleBrokenPage - helper to handle errors
 * @param stopLoading - helper to stop loading
 * @param startLoading - helper to start loading
 * @param tree - the tree if it exists, used to know if we should set loading to true.
 */
export function useGetRootJurisdictionId(
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  handleBrokenPage: HandleBrokenPage,
  stopLoading: StopLoading,
  startLoading: StartLoading,
  tree?: TreeNode
) {
  const [rootJurisdictionId, setRootJurisdictionId] = React.useState<string>('');
  React.useEffect(() => {
    if (plan) {
      const oneOfJurisdictions = plan.jurisdiction.map(
        jurisdictionCode => jurisdictionCode.code
      )[0];
      // get the rootJurisdiction
      const rootJurisdictionKey = 'rootJurisdiction';
      startLoading(rootJurisdictionKey, !tree);
      loadJurisdiction(oneOfJurisdictions, serviceClass)
        .then(result => {
          if (!result || result.error) {
            stopLoading(rootJurisdictionKey);
            throw new Error(COULD_NOT_LOAD_JURISDICTION);
          }
          if (result.value) {
            // TODO: review this - we already have the entire hierarchy in
            // the hierarchy reducer module, do we need to always call OpenSRP?
            // -> this is where we add the hierarchy. we are getting a single jurisdiction from the plan
            // using that to get the root jurisdiction, and then requesting the hierarchy to save to store.
            getAncestors(result.value)
              .then(ancestors => {
                if (ancestors.value) {
                  // get the first ancestor
                  const rootJurisdiction = ancestors.value[0];
                  setRootJurisdictionId(rootJurisdiction.id);
                  stopLoading(rootJurisdictionKey);
                } else {
                  throw new Error(COULD_NOT_LOAD_JURISDICTION);
                }
              })
              .catch(e => {
                return e;
              });
          }
        })
        .catch(error => {
          handleBrokenPage(error.message);
        });
    }
  }, [plan]);

  return rootJurisdictionId;
}

/** get the jurisdiction Tree given the rootJurisdiction Id
 * @param rootJurisdictionId - id of top level jurisdiction
 * @param startLoading - helper to start loading
 * @param treeFetchedCreator - action creator to be called with tree payload once fetched
 * @param stopLoading - helper to stop loading
 * @param handleBrokenPage - helper to handle page errors
 * @param tree - the tree if it exists, used to know if we should set loading to true.
 */
export function useGetJurisdictionTree(
  rootJurisdictionId: string,
  startLoading: StartLoading,
  treeFetchedCreator: ActionCreator<FetchedTreeAction>,
  stopLoading: StopLoading,
  handleBrokenPage: HandleBrokenPage,
  tree?: TreeNode
) {
  React.useEffect(() => {
    const params = {
      return_structure_count: true,
    };
    if (rootJurisdictionId) {
      startLoading(rootJurisdictionId, !tree);
      LoadOpenSRPHierarchy(rootJurisdictionId, OpenSRPService, params)
        .then((apiResponse: Result<RawOpenSRPHierarchy>) => {
          if (apiResponse.value) {
            const responseData = apiResponse.value;
            treeFetchedCreator(responseData);
            stopLoading(rootJurisdictionId);
          }
          if (apiResponse.error) {
            throw new Error(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          }
        })
        .catch(() => {
          handleBrokenPage(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          stopLoading(rootJurisdictionId);
        });
    }
  }, [rootJurisdictionId]);
}
