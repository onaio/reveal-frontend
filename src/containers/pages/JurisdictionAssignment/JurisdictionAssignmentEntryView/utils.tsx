import { Result } from '@onaio/utils/dist/types/types';
import React from 'react';
import { ActionCreator } from 'redux';
import { getAncestors } from '../../../../components/TreeWalker/helpers';
import {
  COULD_NOT_LOAD_JURISDICTION,
  COULD_NOT_LOAD_JURISDICTION_HIERARCHY,
  COULD_NOT_LOAD_PLAN,
} from '../../../../configs/lang';
import { PlanDefinition } from '../../../../configs/settings';
import {
  loadJurisdiction,
  LoadOpenSRPHierarchy,
} from '../../../../helpers/dataLoading/jurisdictions';
import { loadOpenSRPPlan } from '../../../../helpers/dataLoading/plans';
import { OpenSRPService } from '../../../../services/opensrp';
import { RawOpenSRPHierarchy, TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { AddPlanDefinitionAction } from '../../../../store/ducks/opensrp/PlanDefinition';

export function usePlanEffect(
  routePlanId: string,
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  fetchPlanCreator: ActionCreator<AddPlanDefinitionAction>,
  handleBrokenPage: any,
  stopLoading: any,
  startLoading: any
) {
  React.useEffect(() => {
    // create promise to fetch the plan
    const planAndMetaLoadingKey = 'planMetadata';
    startLoading(planAndMetaLoadingKey, !plan);
    const planDataPromise = loadOpenSRPPlan(routePlanId, serviceClass, fetchPlanCreator).catch(
      _ => {
        handleBrokenPage(COULD_NOT_LOAD_PLAN);
      }
    );

    Promise.all([planDataPromise])
      .finally(() => {
        stopLoading(planAndMetaLoadingKey);
      })
      .catch(e => handleBrokenPage(e.message));
  }, []);
}

export function useGetRootJurisdictionId(
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  handleBrokenPage: any,
  stopLoading: any,
  startLoading: any,
  tree?: TreeNode
) {
  const [rootJurisdictionId, setRootJurisdictionId] = React.useState<string>('');
  React.useEffect(() => {
    if (plan) {
      const oneOfJurisdictions = plan.jurisdiction.map(
        jurisdictionCode => jurisdictionCode.code
      )[0];
      // get the rootJurisdiction
      const rootJurisdictionKey = 'rootJurisdiction';
      startLoading(rootJurisdictionKey, !tree);
      loadJurisdiction(oneOfJurisdictions, serviceClass)
        .then(result => {
          if (!result || result.error) {
            stopLoading(rootJurisdictionKey);
            throw new Error(COULD_NOT_LOAD_JURISDICTION);
          }
          if (result.value) {
            // TODO: review this - we already have the entire hierarchy in
            // the hierarchy reducer module, do we need to always call OpenSRP?
            // -> this is where we add the hierarchy. we are getting a single jurisdiction from the plan
            // using that to get the root jurisdiction, and then requesting the hierarchy to save to store.
            getAncestors(result.value)
              .then(ancestors => {
                if (ancestors.value) {
                  // get the first ancestor
                  const rootJurisdiction = ancestors.value[0];
                  setRootJurisdictionId(rootJurisdiction.id);
                  stopLoading(rootJurisdictionKey);
                } else {
                  throw new Error(COULD_NOT_LOAD_JURISDICTION);
                }
              })
              .catch(e => {
                return e;
              });
          }
        })
        .catch(error => {
          handleBrokenPage(error.message);
        });
    }
  }, [plan]);

  return rootJurisdictionId;
}

export function useGetJurisdictionTree(
  rootJurisdictionId: string,
  startLoading: any,
  treeFetchedCreator: any,
  stopLoading: any,
  handleBrokenPage: any,
  tree?: TreeNode
) {
  React.useEffect(() => {
    const params = {
      return_structure_count: true,
    };
    if (rootJurisdictionId) {
      startLoading(rootJurisdictionId, !tree);
      LoadOpenSRPHierarchy(rootJurisdictionId, OpenSRPService, params)
        .then((apiResponse: Result<RawOpenSRPHierarchy>) => {
          if (apiResponse.value) {
            const responseData = apiResponse.value;
            treeFetchedCreator(responseData);
            stopLoading(rootJurisdictionId);
          }
          if (apiResponse.error) {
            throw new Error(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          }
        })
        .catch(() => {
          handleBrokenPage(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          stopLoading(rootJurisdictionId);
        });
    }
  }, [rootJurisdictionId]);
}
