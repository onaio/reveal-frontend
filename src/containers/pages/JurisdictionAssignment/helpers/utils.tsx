import { useState } from 'react';
import { displayError } from '../../../../helpers/errors';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/** tells whether we should select the parent checkbox in the table header
 * @param currentParentNode - the current parent node
 * @param currentChildren - current children
 */
export const checkParentCheckbox = (
  currentParentNode: TreeNode | undefined,
  currentChildren: TreeNode[]
) => {
  let selected = true;
  if (currentParentNode) {
    selected = selected && nodeIsSelected(currentParentNode);
  } else {
    // when currentParent node is undefined but there are currentChildren usually at the top level
    if (currentChildren.length > 0) {
      currentChildren.forEach(node => {
        selected = selected && nodeIsSelected(node);
      });
    } else {
      selected = false;
    }
  }
  return selected;
};

export const useHandleBrokenPage = () => {
  const [broken, setBroken] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string | undefined>();

  /** Convenience function to handle cases where we must abort and tell the user we have done so */
  const handleBrokenPage = (message: string) => {
    displayError(Error(message));
    setErrorMessage(message);
    setBroken(true);
  };

  return { broken, errorMessage, handleBrokenPage };
};
