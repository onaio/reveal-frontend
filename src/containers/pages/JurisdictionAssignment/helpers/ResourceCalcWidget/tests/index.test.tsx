import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { cloneDeep } from 'lodash';
import React from 'react';
import { Provider } from 'react-redux';
import { computeEstimate, ConnectedResourceWidget, ResourceCalculation } from '..';
import store from '../../../../../../store';
import { fetchTree, selectNode } from '../../../../../../store/ducks/opensrp/hierarchies';
import { META_STRUCTURE_COUNT } from '../../../../../../store/ducks/opensrp/hierarchies/constants';
import { raZambiaHierarchy } from '../../../../../../store/ducks/opensrp/hierarchies/tests/fixtures';
import { generateJurisdictionTree } from '../../../../../../store/ducks/opensrp/hierarchies/utils';

const rootId = '0ddd9ad1-452b-4825-a92a-49cb9fc82d18';
const planId = 'randomPlanId';
store.dispatch(fetchTree(raZambiaHierarchy));
store.dispatch(selectNode(rootId, rootId, planId));

const rootNode = generateJurisdictionTree(raZambiaHierarchy);

const props = {
  planId,
  rootId,
};

describe('jurisdiction Assignment/Resource calculation', () => {
  it('renders without crashing', () => {
    shallow(<ResourceCalculation {...props} />);
  });

  it('renders correctly', () => {
    const currentParentNode = cloneDeep(rootNode);
    currentParentNode.model.meta[META_STRUCTURE_COUNT] = 908;
    const otherProps = {
      ...props,
      selectedTreeCurrentParentNode: currentParentNode,
    };
    const wrapper = mount(<ResourceCalculation {...otherProps} />);

    expect(wrapper.find('h3.section-title').text()).toMatchInlineSnapshot(
      `"Resource Estimate for ra Zambia"`
    );

    expect(wrapper.find('form').text()).toMatchInlineSnapshot(
      `"0 daysat a rate of   structures per team per day with  Teams"`
    );
  });

  it('computes the time estimates correctly', () => {
    // need to compute for nominal case
    let res = computeEstimate(30, 10, 3);
    expect(res).toEqual(1);

    // fraction days
    res = computeEstimate(31, 10, 3);
    expect(res).toEqual(2);

    // division by zero error
    res = computeEstimate(3, 0, 2);
    expect(res).toEqual(0);
    res = computeEstimate(3, 10, 0);
    expect(res).toEqual(0);
  });

  it('uses user values to compute time estimates correctly', () => {
    // 2 cases:
    // - there is data.
    // user backspaces such that there is no data
    const currentParentNode = cloneDeep(rootNode);
    currentParentNode.model.meta[META_STRUCTURE_COUNT] = 30;
    const otherProps = {
      ...props,
      selectedTreeCurrentParentNode: currentParentNode,
    };
    const wrapper = mount(<ResourceCalculation {...otherProps} />);

    // initial input values
    expect(wrapper.find('input[name="structuresCount"]').props().value).toEqual(0);
    expect(wrapper.find('input[name="teamsCount"]').props().value).toEqual(0);

    // simulate structure count change so that we can have 10 structures per team
    expect(toJson(wrapper.find('input[name="structuresCount"]'))).toMatchSnapshot(
      'structures count input'
    );
    wrapper
      .find('input[name="structuresCount"]')
      .simulate('change', { target: { value: 10, name: 'structuresCount' } });
    wrapper.update();

    expect(wrapper.find('input[name="structuresCount"]').props().value).toEqual(10);
    expect(wrapper.find('input[name="teamsCount"]').props().value).toEqual(0);

    // see if there has been any change to the number of days, should start with 0 days
    expect(wrapper.find('form p').text()).toMatchInlineSnapshot(
      `"0 daysat a rate of   structures per team per day with  Teams"`
    );

    // simulate teams count change so that we have 3 teams
    expect(toJson(wrapper.find('input[name="teamsCount"]'))).toMatchSnapshot(
      'structures count input'
    );
    wrapper
      .find('input[name="teamsCount"]')
      .simulate('change', { target: { value: 3, name: 'teamsCount' } });
    wrapper.update();

    expect(wrapper.find('input[name="structuresCount"]').props().value).toEqual(10);
    expect(wrapper.find('input[name="teamsCount"]').props().value).toEqual(3);

    // see if there has been any change to the number of days, should start with 1 days
    expect(wrapper.find('form p').text()).toMatchInlineSnapshot(
      `"1 daysat a rate of   structures per team per day with  Teams"`
    );

    // backspace teams so that we have no data, value would be ""
    wrapper
      .find('input[name="teamsCount"]')
      .simulate('change', { target: { value: '', name: 'teamsCount' } });
    wrapper.update();

    expect(wrapper.find('input[name="structuresCount"]').props().value).toEqual(10);
    expect(wrapper.find('input[name="teamsCount"]').props().value).toEqual('');

    // see if there has been any change to the number of days, should start with 0 days
    expect(wrapper.find('form p').text()).toMatchInlineSnapshot(
      `"0 daysat a rate of   structures per team per day with  Teams"`
    );
  });

  it('works with store', () => {
    const otherProps = {
      ...props,
      currentParentId: rootId,
    };
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedResourceWidget {...otherProps} />
      </Provider>
    );

    expect(wrapper.text().includes(rootNode.model.label)).toBeTruthy();
  });
});
