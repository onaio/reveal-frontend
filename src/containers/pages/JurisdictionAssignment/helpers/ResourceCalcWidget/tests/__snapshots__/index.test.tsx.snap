// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jurisdiction Assignment/Resource calculation uses user values to compute time estimates correctly: structures count input 1`] = `
<input
  name="structuresCount"
  onBlur={[Function]}
  onChange={[Function]}
  type="number"
  value={0}
/>
`;

exports[`jurisdiction Assignment/Resource calculation uses user values to compute time estimates correctly: structures count input 2`] = `
<input
  name="teamsCount"
  onBlur={[Function]}
  onChange={[Function]}
  type="number"
  value={0}
/>
`;
